File: aws_batch_squared_py/nextflow_stack.py
"""Nextflow nested stack — head-node job role and job definition.

Ported from ``lib/nextflow-stack.ts``. The IAM policies and container env vars
reference SSM parameters via CloudFormation ``{{resolve:ssm:...}}`` dynamic
references so queue ARNs and the bucket name are wired at deploy time.
"""

from typing import Optional

import aws_cdk as cdk
from aws_cdk import aws_batch as batch
from aws_cdk import aws_iam as iam
from constructs import Construct


class NextflowStack(cdk.NestedStack):
    nextflow_job_role: iam.Role
    nextflow_job_definition: batch.CfnJobDefinition

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        namespace: str,
        group_name: str,
        nextflow_image: str,
        s3_bucket_param: str,
        on_demand_queue_param: str,
        spot_queue_param: str,
        s3_nextflow_prefix: Optional[str] = None,
        s3_logs_dir_prefix: Optional[str] = None,
        s3_work_dir_prefix: Optional[str] = None,
        **kwargs,
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        account = cdk.Stack.of(self).account
        region = cdk.Stack.of(self).region

        s3_nextflow_prefix = s3_nextflow_prefix or "_nextflow"
        s3_logs_dir_prefix = s3_logs_dir_prefix or "logs"
        s3_work_dir_prefix = s3_work_dir_prefix or "runs"

        # Create Nextflow Job Role
        self.nextflow_job_role = iam.Role(
            self,
            "IAMNextflowJobRole",
            assumed_by=iam.ServicePrincipal("ecs-tasks.amazonaws.com"),
            managed_policies=[
                iam.ManagedPolicy.from_aws_managed_policy_name("AmazonS3ReadOnlyAccess"),
            ],
            inline_policies={
                f"Nextflow-Batch-Access-{region}": iam.PolicyDocument(
                    statements=[
                        # Batch read access
                        iam.PolicyStatement(
                            sid="BatchReadAccessAllowAll",
                            effect=iam.Effect.ALLOW,
                            resources=["*"],
                            actions=["batch:List*", "batch:Describe*"],
                        ),
                        # Batch tagging permissions
                        iam.PolicyStatement(
                            sid="BatchTagAccess",
                            effect=iam.Effect.ALLOW,
                            resources=["*"],
                            actions=[
                                "batch:TagResource",
                                "batch:ListTagsForResource",
                                "batch:UntagResource",
                            ],
                        ),
                        # CloudWatch logs read access
                        iam.PolicyStatement(
                            sid="CloudwatchReadLogEvents",
                            effect=iam.Effect.ALLOW,
                            resources=[
                                f"arn:aws:logs:{region}:{account}:log-group:"
                                "/aws/batch/job:log-stream:*"
                            ],
                            actions=["logs:GetLogEvents"],
                        ),
                        # Batch job submission
                        iam.PolicyStatement(
                            sid="BatchWriteAccessAllowJobSubmission",
                            effect=iam.Effect.ALLOW,
                            resources=[
                                f"{{{{resolve:ssm:{on_demand_queue_param}:1}}}}",
                                f"{{{{resolve:ssm:{spot_queue_param}:1}}}}",
                                "arn:aws:batch:*:*:job-definition/nf-*:*",
                            ],
                            actions=["batch:*Job"],
                        ),
                        # Batch job definition management
                        iam.PolicyStatement(
                            sid="BatchWriteAccessAllowJobDefinition",
                            effect=iam.Effect.ALLOW,
                            resources=[
                                "arn:aws:batch:*:*:job-definition/nf-*",
                                "arn:aws:batch:*:*:job-definition/nf-*:*",
                            ],
                            actions=["batch:*JobDefinition"],
                        ),
                    ]
                ),
                f"Nextflow-S3Bucket-Access-{region}": iam.PolicyDocument(
                    statements=[
                        iam.PolicyStatement(
                            effect=iam.Effect.ALLOW,
                            resources=[
                                f"arn:aws:s3:::{{{{resolve:ssm:{s3_bucket_param}:1}}}}",
                                f"arn:aws:s3:::{{{{resolve:ssm:{s3_bucket_param}:1}}}}/*",
                            ],
                            actions=["s3:*"],
                        ),
                    ]
                ),
                f"Nextflow-Instance-Access-{region}": iam.PolicyDocument(
                    statements=[
                        iam.PolicyStatement(
                            effect=iam.Effect.ALLOW,
                            actions=[
                                "ecs:DescribeTasks",
                                "ec2:DescribeInstances",
                                "ec2:DescribeInstanceTypes",
                                "ec2:DescribeInstanceAttribute",
                                "ecs:DescribeContainerInstances",
                                "ec2:DescribeInstanceStatus",
                            ],
                            resources=["*"],
                        ),
                    ]
                ),
            },
        )

        # Create Nextflow Job Definition
        self.nextflow_job_definition = batch.CfnJobDefinition(
            self,
            "BatchNextflowJobDefinition",
            job_definition_name=f"nextflow-{namespace}",
            type="container",
            timeout=batch.CfnJobDefinition.TimeoutProperty(
                attempt_duration_seconds=3600,
            ),
            container_properties=batch.CfnJobDefinition.ContainerPropertiesProperty(
                memory=16384,
                vcpus=4,
                image=nextflow_image,
                job_role_arn=self.nextflow_job_role.role_arn,
                mount_points=[
                    batch.CfnJobDefinition.MountPointsProperty(
                        source_volume="aws-cli",
                        container_path="/opt/aws-cli",
                        read_only=True,
                    )
                ],
                volumes=[
                    batch.CfnJobDefinition.VolumesProperty(
                        name="aws-cli",
                        host=batch.CfnJobDefinition.VolumesHostProperty(
                            source_path="/opt/aws-cli",
                        ),
                    )
                ],
                environment=[
                    # Work (child) jobs default to the Spot queue for cost savings; they are
                    # retriable so Spot interruptions are tolerable. The head node itself still
                    # runs on the On-Demand queue (that's the queue you submit this job to).
                    # Override per-run by setting NF_JOB_QUEUE in the submit-job container env.
                    batch.CfnJobDefinition.EnvironmentProperty(
                        name="NF_JOB_QUEUE",
                        value=f"{{{{resolve:ssm:{spot_queue_param}:1}}}}",
                    ),
                    batch.CfnJobDefinition.EnvironmentProperty(
                        name="NF_LOGSDIR",
                        value=(
                            f"s3://{{{{resolve:ssm:{s3_bucket_param}:1}}}}/"
                            f"{s3_nextflow_prefix}/{s3_logs_dir_prefix}"
                        ),
                    ),
                    batch.CfnJobDefinition.EnvironmentProperty(
                        name="NF_WORKDIR",
                        value=(
                            f"s3://{{{{resolve:ssm:{s3_bucket_param}:1}}}}/"
                            f"{s3_nextflow_prefix}/{s3_work_dir_prefix}"
                        ),
                    ),
                ],
            ),
        )

        # Outputs
        cdk.CfnOutput(
            self,
            "NextflowBucket",
            value=f"s3://{{{{resolve:ssm:{s3_bucket_param}:1}}}}",
            description=(
                "S3 Bucket used to store Nextflow metadata (session cache, logs, and "
                "intermediate results)"
            ),
        )
        cdk.CfnOutput(
            self,
            "LogsDir",
            value=(
                f"s3://{{{{resolve:ssm:{s3_bucket_param}:1}}}}/"
                f"{s3_nextflow_prefix}/{s3_logs_dir_prefix}"
            ),
            description="S3 URI where nextflow session cache and logs are stored",
        )
        cdk.CfnOutput(
            self,
            "WorkDir",
            value=(
                f"s3://{{{{resolve:ssm:{s3_bucket_param}:1}}}}/"
                f"{s3_nextflow_prefix}/{s3_work_dir_prefix}"
            ),
            description="S3 URI where workflow intermediate results are stored",
        )
        cdk.CfnOutput(
            self,
            "NextflowJobDefinition",
            value=self.nextflow_job_definition.ref,
            description=(
                "Batch Job Definition that creates a nextflow head node for running "
                "workflows"
            ),
        )
        cdk.CfnOutput(
            self,
            "NextflowJobRole",
            value=self.nextflow_job_role.role_arn,
            description=(
                "IAM Role that allows the nextflow head node job access to S3 and Batch"
            ),
        )

File: playbook/ways-to-run/patterns/batch-squared/infrastructure/lib/nextflow-stack.ts
import * as cdk from "aws-cdk-lib";
import * as batch from "aws-cdk-lib/aws-batch";
import * as iam from "aws-cdk-lib/aws-iam";
import type { Construct } from "constructs";

export interface NextflowStackProps extends cdk.NestedStackProps {
  namespace: string;
  groupName: string;
  nextflowImage: string;
  s3BucketParam: string;
  onDemandQueueParam: string;
  spotQueueParam: string;
  s3NextflowPrefix?: string;
  s3LogsDirPrefix?: string;
  s3WorkDirPrefix?: string;
}

export class NextflowStack extends cdk.NestedStack {
  public readonly nextflowJobRole: iam.Role;
  public readonly nextflowJobDefinition: batch.CfnJobDefinition;

  constructor(scope: Construct, id: string, props: NextflowStackProps) {
    super(scope, id, props);

    const account = cdk.Stack.of(this).account;
    const region = cdk.Stack.of(this).region;

    const s3NextflowPrefix = props.s3NextflowPrefix || "_nextflow";
    const s3LogsDirPrefix = props.s3LogsDirPrefix || "logs";
    const s3WorkDirPrefix = props.s3WorkDirPrefix || "runs";

    // Create Nextflow Job Role
    this.nextflowJobRole = new iam.Role(this, "IAMNextflowJobRole", {
      assumedBy: new iam.ServicePrincipal("ecs-tasks.amazonaws.com"),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName("AmazonS3ReadOnlyAccess"),
      ],
      inlinePolicies: {
        [`Nextflow-Batch-Access-${region}`]: new iam.PolicyDocument({
          statements: [
            // Batch read access
            new iam.PolicyStatement({
              sid: "BatchReadAccessAllowAll",
              effect: iam.Effect.ALLOW,
              resources: ["*"],
              actions: ["batch:List*", "batch:Describe*"],
            }),
            // Batch tagging permissions
            new iam.PolicyStatement({
              sid: "BatchTagAccess",
              effect: iam.Effect.ALLOW,
              resources: ["*"],
              actions: [
                "batch:TagResource",
                "batch:ListTagsForResource",
                "batch:UntagResource",
              ],
            }),
            // CloudWatch logs read access
            new iam.PolicyStatement({
              sid: "CloudwatchReadLogEvents",
              effect: iam.Effect.ALLOW,
              resources: [
                `arn:aws:logs:${region}:${account}:log-group:/aws/batch/job:log-stream:*`,
              ],
              actions: ["logs:GetLogEvents"],
            }),
            // Batch job submission
            new iam.PolicyStatement({
              sid: "BatchWriteAccessAllowJobSubmission",
              effect: iam.Effect.ALLOW,
              resources: [
                `{{resolve:ssm:${props.onDemandQueueParam}:1}}`,
                `{{resolve:ssm:${props.spotQueueParam}:1}}`,
                "arn:aws:batch:*:*:job-definition/nf-*:*",
              ],
              actions: ["batch:*Job"],
            }),
            // Batch job definition management
            new iam.PolicyStatement({
              sid: "BatchWriteAccessAllowJobDefinition",
              effect: iam.Effect.ALLOW,
              resources: [
                "arn:aws:batch:*:*:job-definition/nf-*",
                "arn:aws:batch:*:*:job-definition/nf-*:*",
              ],
              actions: ["batch:*JobDefinition"],
            }),
          ],
        }),
        [`Nextflow-S3Bucket-Access-${region}`]: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              resources: [
                `arn:aws:s3:::{{resolve:ssm:${props.s3BucketParam}:1}}`,
                `arn:aws:s3:::{{resolve:ssm:${props.s3BucketParam}:1}}/*`,
              ],
              actions: ["s3:*"],
            }),
          ],
        }),
        [`Nextflow-Instance-Access-${region}`]: new iam.PolicyDocument({
          statements: [
            new iam.PolicyStatement({
              effect: iam.Effect.ALLOW,
              actions: [
                "ecs:DescribeTasks",
                "ec2:DescribeInstances",
                "ec2:DescribeInstanceTypes",
                "ec2:DescribeInstanceAttribute",
                "ecs:DescribeContainerInstances",
                "ec2:DescribeInstanceStatus",
              ],
              resources: ["*"],
            }),
          ],
        }),
      },
    });

    // Create Nextflow Job Definition
    this.nextflowJobDefinition = new batch.CfnJobDefinition(
      this,
      "BatchNextflowJobDefinition",
      {
        jobDefinitionName: `nextflow-${props.namespace}`,
        type: "container",
        timeout: {
          attemptDurationSeconds: 3600,
        },
        containerProperties: {
          memory: 16384,
          vcpus: 4,
          image: props.nextflowImage,
          jobRoleArn: this.nextflowJobRole.roleArn,
          mountPoints: [
            {
              sourceVolume: "aws-cli",
              containerPath: "/opt/aws-cli",
              readOnly: true,
            },
          ],
          volumes: [
            {
              name: "aws-cli",
              host: {
                sourcePath: "/opt/aws-cli",
              },
            },
          ],
          environment: [
            {
              // Work (child) jobs default to the Spot queue for cost savings; they are
              // retriable so Spot interruptions are tolerable. The head node itself still
              // runs on the On-Demand queue (that's the queue you submit this job to).
              // Override per-run by setting NF_JOB_QUEUE in the submit-job container env.
              name: "NF_JOB_QUEUE",
              value: `{{resolve:ssm:${props.spotQueueParam}:1}}`,
            },
            {
              name: "NF_LOGSDIR",
              value: `s3://{{resolve:ssm:${props.s3BucketParam}:1}}/${s3NextflowPrefix}/${s3LogsDirPrefix}`,
            },
            {
              name: "NF_WORKDIR",
              value: `s3://{{resolve:ssm:${props.s3BucketParam}:1}}/${s3NextflowPrefix}/${s3WorkDirPrefix}`,
            },
          ],
        },
      },
    );

    // Outputs
    new cdk.CfnOutput(this, "NextflowBucket", {
      value: `s3://{{resolve:ssm:${props.s3BucketParam}:1}}`,
      description:
        "S3 Bucket used to store Nextflow metadata (session cache, logs, and intermediate results)",
    });

    new cdk.CfnOutput(this, "LogsDir", {
      value: `s3://{{resolve:ssm:${props.s3BucketParam}:1}}/${s3NextflowPrefix}/${s3LogsDirPrefix}`,
      description: "S3 URI where nextflow session cache and logs are stored",
    });

    new cdk.CfnOutput(this, "WorkDir", {
      value: `s3://{{resolve:ssm:${props.s3BucketParam}:1}}/${s3NextflowPrefix}/${s3WorkDirPrefix}`,
      description: "S3 URI where workflow intermediate results are stored",
    });

    new cdk.CfnOutput(this, "NextflowJobDefinition", {
      value: this.nextflowJobDefinition.ref,
      description:
        "Batch Job Definition that creates a nextflow head node for running workflows",
    });

    new cdk.CfnOutput(this, "NextflowJobRole", {
      value: this.nextflowJobRole.roleArn,
      description:
        "IAM Role that allows the nextflow head node job access to S3 and Batch",
    });
  }
}
